bootstrap/internal/controller/infraenv_controller.go

```diff
@@ -18,23 +18,39 @@ package controller
 
 import (
 	"context"
+	"fmt"
+	"net/url"
 
 	bootstrapv1alpha1 "github.com/openshift-assisted/cluster-api-agent/bootstrap/api/v1alpha1"
 	logutil "github.com/openshift-assisted/cluster-api-agent/util/log"
 	aiv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
 
 	"github.com/pkg/errors"
 
+	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type InfraEnvControllerConfig struct {
+	// UseInternalImageURL when set to false means that we'll use the InfraEnv's iso download URL
+	// as is. When set to true, it'll use the assisted-image-service's internal IP as part of the
+	// download URL.
+	UseInternalImageURL bool `envconfig:"USE_INTERNAL_IMAGE_URL" default:"true"`
+	// ImageServiceName is the Service CR name for the assisted-image-service
+	ImageServiceName string `envconfig:"IMAGE_SERVICE_NAME" default:"assisted-image-service"`
+	// ImageServiceNamespace is the namespace that the Service CR for the assisted-image-service is in
+	ImageServiceNamespace string `envconfig:"IMAGE_SERVICE_NAMESPACE" default:"assisted-installer"`
```

Comment:
Should this have a default from within the code? Normally we deploy this way, but can we really assume it from here? Wouldn't it be safer to force the use specify this?

Reply:
This is part of the reason I asked @jianzzha to review.

I don't know what we will "normally" be doing.
When assisted is deployed through MCE it won't be in the `assisted-installer` namespace, but when we deploy it for dev, it is. I'm also not sure what namespace whatever MCE distribution we use for sylva will use. I don't know what to consider "normal" in this case.?

Reply:
Sylva is referencing  https://github.com/openshift/assisted-service/tree/master/config/default kustomize folder at this moment, so  'assisted-installer' namespace is set by that kustomize. 

I'm assuming that MCE can kustomize the IMAGE_SERVICE_NAMESPACE if it needs to?

Reply:
> I'm assuming that MCE can kustomize the IMAGE_SERVICE_NAMESPACE if it needs to?

This is what I'm trying to ensure. MCE does not use kustomize when deploying assisted. I assume it also won't use kustomize when deploying this CAPI provider (if it is even going to be responsible for that). MCE pulls the files from the rendered operator bundle and applies those.

Reply:
But the open question (for me at least) is still if MCE is going to deploy the CAPI providers or if they're being installed through some other means.

Reply:
When MCE deploys this we'll be under OCP, so public route can be used, thus turning off this hack
I think independently of what we would "normally" do, the safer option is to default to public route (advertised), and only when specifically overriding it it should be set. Of course we should reflect this in the docs as suggested.
We could also have a sample overlay already prepared where users could apply it with current default (basically moving defaults from code to a sample kustomize directory).

Reply:
>When MCE deploys this we'll be under OCP

No? Wasn't part of this effort to get MCE to install on vanilla kube as well?

Reply:
Discussed offline with Nick that we should just look for assisted service in the same namespace as where the bootstrap controller is running since we have control over where this controller runs. 

In other words, we'll deploy this controller in the same namespace that assisted service runs.

Reply:
This is fine. We have a kustomize sylva unit for both the infra operator and capi provider and can make sure they run in the same name space.

Reply:
> In other words, we'll deploy this controller in the same namespace that assisted service runs.
I am not sure I agree with this. What is the advantage over the other possible solutions?

I can see the following scenarios:
* when override behaviour is disabled, we use public route: no need to know where assisted is runnins
* when override behaviour is enabled, we could do one of the following:
  * the user needs to specify where assisted is running
  * the user must deploy assisted and CAPI controllers in the same namespace

I think the current proposed default it's the least flexible scenario.


The advantages I see as leaving as default the public route is that:
* this should be the desired behaviour: reading from advertised address. OCP will work this way, and I believe if this is the way to go for the future, we should eventually advertise the private address. The current solution is a workaround.
 
We can always override and eventually stop overriding from Sylva unit.

Is there something that I am missing here?

Reply:
I suggested looking in the same namespace as a more sensible default than hardcoding `assisted-installer`.

I think we should still expose the env var as overrideable, but it was my understanding that we didn't have a good way to do that from kustomize without hardcoding other namespaces.

As for where these components are running we only have three cases:
1. Everything installed from MCE -> everything runs in the same namespace
2. Assisted installed from MCE, CAPI provider installed by user -> doc to run CAPI provider in assisted namespace
3. User installs everything -> doc to run everything in the same namespace.

From what I could tell from the discussion around kustomize overrides for the namespace this seems more reliable.

Reply:
I see and agree that running everything in the same namespace might be more sensible than hardcoding assisted's namespace. What about the following:
* Default to USE_INTERNAL_IMAGE_URL to false: this will use public route, where all of this is a non-issue and it would be the desired behaviour
* Default IMAGE_SERVICE_NAMESPACE to empty: when empty it will mean same namespace (if override is turned on), however it can be overridden if need be.

In Sylva's "upstream" case, we'd only need to add one env var through kustomization (USE_INTERNAL_IMAGE_URL=true) and deploy in the same namespace.
In MCE case, everything would be deployed in the same namespace, but no overrides needed as we'd go through the advertised public address.

Basically my point is that the "workaround" (USE_INTERNAL_IMAGE_URL) should not be default behaviour. Would you agree?

Reply:
>In MCE case, everything would be deployed in the same namespace, but no overrides needed

You're still assuming whenever we run with MCE we're going to be in OCP and I'm not sure yet that this will always be true.
But that said, I beleive MCE's installer can set env vars when it deploys operators so we'd have to get them to set this in the OCP/kube case no matter what we choose so I don't think it's an issue.

Plus we're still a bit away from having MCE install this provider so we can probably deal with this when it's actually an issue.

Reply:
Wouldn't it be easier to just remove any "workaround" kustomization rather than having to change the upstream manifest, when/if we'll finally release that?
I feel we are making a workaround the default behaviour making some assumptions (what namespace we run what component) on the way. Shouldn't we allow the user to have this workaround and make the assumptions/decisions themselves?

Reply:
Had a discussion offline, the decision was to remove the default value for the image service namespace. If provided we will use it to look up the service, if not we will look for the service in the namespace the provider is running in.

We will also default `USE_INTERNAL_IMAGE_URL` to `false` and the user will need to change that if they want/need to.